walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,151 @@
+    // return those time slots. 
+    if (!timeRangesForAllAttendees.isEmpty()) {
+      return timeRangesForAllAttendees;
+    }
+    
+    // Otherwise, return the time slots that fit just the mandatory attendees.
+    return solveQueryforAttendees(eventsSet, mandatoryAttendees, request.getDuration());
+  }
+
+  /**
+   * Return the times when a new meeting could happen in a day, so that all {@code attendees} can join.
+   *
+   * @param events    The set of all known events in that day.
+   * @param attendees The set of required attendees for the new meeting.
+   * @param duration  The time duration of the new meeting.
+   */
+  private static List<TimeRange> solveQueryforAttendees(Set<Event> events, Set<String> attendees, long duration) {
+    // Get the time ranges when at least one attendee is not free 
+    // (because for a particular time slot to work, all attendees must be free to attend the new meeting).
+    List<TimeRange> occupiedTimeRanges = events
+      .stream()
+      .filter(event -> FindMeetingQuery.hasCommonAttendees(event, attendees))
+      .map(event -> event.getWhen())
+      .sorted(TimeRange.ORDER_BY_START)
+      .collect(Collectors.toList());
+    
+    List<TimeRange> result = getNewMeetingTimeRanges(occupiedTimeRanges, duration);
+
+    return result;
+  }
+
+  /**
+   * Returns whether the {@code event} has at least one common attendee as the {@code requestedAttendees}.
+   */
+  private static boolean hasCommonAttendees(Event event, Set<String> requestedAttendees) {
+    Set<String> eventAttendees = new HashSet<>(event.getAttendees());
+    for (String requestedAttendee : requestedAttendees) {
+      if (eventAttendees.contains(requestedAttendee)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+
+  /**
+   * Returns the times slots when a new meeting taking {@code newMeetingDuration} minutes could happen 
+   * in a day, considering the {@code occupiedTimeRanges}, which are sorted in ascending order by start.
+   * It uses a Greedy algorithm.
+   */
+  private static List<TimeRange> getNewMeetingTimeRanges(List<TimeRange> occupiedTimeRanges, 
+                                                         long newMeetingDuration) {
+    // Consider the cases when there are no constraints for the new meeting.
+    if (occupiedTimeRanges.isEmpty()) {
+      if (newMeetingDuration <= TimeRange.WHOLE_DAY.duration()) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+      } else {
+        return Collections.emptyList();
+      }
+    }
+
+    // Init the time slots list when the new meeting can take place. 
+    List<TimeRange> newMeetingTimeRanges = new ArrayList<>();
+
+    Iterator<TimeRange> occupiedTimeRangesIterator = occupiedTimeRanges.iterator();
+    TimeRange currentOccupiedTimeRange = occupiedTimeRangesIterator.next();
+
+    // Check for empty time slot before the first occupied time range
+    if (currentOccupiedTimeRange.start() > TimeRange.START_OF_DAY) {
```

Comment:
This part doesn't make too much sense to me. 

* what will happen if there are several "occupiedTimeRanges" start earlier than "START_OF_DAY"? what if they end before the "START_OF_DAY"?

The logic may be as simple as:

#1 sort occupied time ranges
#2 loop all "space" between time ranges.
#3 for each space, check whether it's within the **day** and longer than the "newMeetingDuration". If yes, add to the result.

Reply:
> * what will happen if there are several "occupiedTimeRanges" start earlier than "START_OF_DAY"? what if they end before the "START_OF_DAY"?

The _"Check for empty time slot before the first occupied time range"_ will still work. My logic is the same as yours. But I still have to do the check for an empty time range "in the beginning of the day", before the first existing event starts and this is what I'm doing there.

What I hadn't yet added, but will add, is checking whether the time range is within the day when Ioop all the spaces between time ranges

Reply:
I filtered the `occupiedTimeRanges` in the `solveQueryforAttendees()` so that all of them are within the day, hope it's fine this way. Now I know that all the time ranges will be within the day.

My logic is the same as yours. But I still have to do the check for an empty time range "in the beginning of the day", before the first existing event starts and this is what I'm doing there.

Reply:
What a good idea (the one with the dummy values)! I made sure all time ranges are within the day by filtering the occupiedTimeRanges, and I added the two dummy values. So, done!